jsonpb/jsonpb_test.go

```diff
@@ -721,6 +721,61 @@ func TestUnmarshalingBadInput(t *testing.T) {
 	}
 }
 
+type funcResolver func(turl string) (proto.Message, error)
+
+func (fn funcResolver) Resolve(turl string) (proto.Message, error) {
+	return fn(turl)
+}
+
+func TestAnyWithCustomResolver(t *testing.T) {
+	resolveCount := 0
+	resolver := funcResolver(func(turl string) (proto.Message, error) {
+		resolveCount++
+		return new(pb.Simple), nil
+	})
+	msg := &pb.Simple{
+		OBytes:  []byte{1, 2, 3, 4},
+		OBool:   proto.Bool(true),
+		OString: proto.String("foobar"),
+		OInt64:  proto.Int64(1020304),
+	}
+	msgBytes, err := proto.Marshal(msg)
+	if err != nil {
+		t.Errorf("an unexpected error occurred when marshaling message: %v", err)
+	}
+	// make an Any with a type URL that won't resolve w/out custom resolver
+	any := &anypb.Any{
+		TypeUrl: "https://foobar.com/some.random.MessageKind",
+		Value:   msgBytes,
+	}
+
+	m := Marshaler{UrlResolver: resolver}
+	js, err := m.MarshalToString(any)
+	if err != nil {
+		t.Errorf("an unexpected error occurred when marshaling any to JSON: %v", err)
+	}
+	if resolveCount != 1 {
+		t.Errorf("custom resolver was not invoked during marshaling")
+	}
+	wanted := `{"@type":"https://foobar.com/some.random.MessageKind","oBool":true,"oInt64":"1020304","oString":"foobar","oBytes":"AQIDBA=="}`
```

Comment:
Shouldn't the JSON keys be sorted?

Reply:
If the object marshaled is a map, then the keys are sorted (since maps have no defined order, but JSON output wants to be deterministic). In this case, however, the object being marshaled is a struct, so it's fields are marshaled in declaration order.

Reply:
Ah, yes.  I got confused looking at code section that marshals via JSONPBMarshaler -- https://github.com/golang/protobuf/blob/master/jsonpb/jsonpb.go#L144.  Thanks for pointing that out.